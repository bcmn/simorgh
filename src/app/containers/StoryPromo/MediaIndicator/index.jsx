import React from 'react';
import moment from 'moment-timezone';
import { shape, bool, string } from 'prop-types';
import MediaIndicatorComp from '@bbc/psammead-media-indicator';
import pathOr from 'ramda/src/pathOr';
import { storyItem } from '../../../models/propTypes/storyItem';
import formatDuration from '../../../lib/utilities/formatDuration';

const MediaIndicator = ({ item, topStory, service }) => {
  const isPGL = pathOr(null, ['cpsType'], item) === 'PGL';
  const isMedia = pathOr(null, ['cpsType'], item) === 'MAP';
  const hasMediaInfo = pathOr(null, ['media'], item);

  // Only build a media indicator if this is a photo gallery or media item
  if (!isPGL && (!isMedia || !hasMediaInfo)) {
    return null;
  }

  const type = isPGL ? 'photogallery' : pathOr(null, ['media', 'format'], item);

  // Always gets the first version. Smarter logic may be needed in the future.
  const rawDuration = pathOr(null, ['media', 'versions', 0, 'duration'], item);

  if (rawDuration) {
    const duration = moment.duration(rawDuration, 'seconds');
    const durationString = formatDuration(duration);
    const isoDuration = duration.toISOString();
    return (
      <MediaIndicatorComp
        duration={durationString}
        datetime={isoDuration}
        type={type}
        topStory={topStory}
        service={service}
      />
    );
  }

  return <MediaIndicatorComp type={type} service={service} />;
};

MediaIndicator.propTypes = {
  item: shape(storyItem).isRequired,
  topStory: bool,
  service: string.isRequired,
};

MediaIndicator.defaultProps = {
  topStory: false,
};

export default MediaIndicator;
