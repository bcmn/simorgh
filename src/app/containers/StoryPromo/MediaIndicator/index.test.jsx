import React from 'react';

import { shouldShallowMatchSnapshot, isNull } from '../../../../testHelpers';
import MediaIndicator from '.';

const audioItem = {
  headlines: {
    headline: 'An audio item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'MAP',
  media: {
    format: 'audio',
    versions: [
      {
        duration: 59,
      },
    ],
  },
};

const videoItem = {
  headlines: {
    headline: 'A video item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'MAP',
  media: {
    format: 'video',
    versions: [
      {
        duration: 5600,
      },
    ],
  },
};

const photogalleryItem = {
  headlines: {
    headline: 'A photo gallery item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'PGL',
};

const nonMediaItem = {
  cpsType: 'ðŸ¤”',
};

const noDurationItem = {
  headlines: {
    headline: 'A video item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'MAP',
  media: {
    format: 'video',
    versions: [{}],
  },
};

const noMediaObject = {
  headlines: {
    headline: 'A video item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'MAP',
};

const noMediaFormat = {
  headlines: {
    headline: 'A video item',
  },
  locators: {
    assetUri: 'https://www.bbc.co.uk',
  },
  cpsType: 'MAP',
  media: {
    versions: [
      {
        duration: 5600,
      },
    ],
  },
};

describe('Story Promo Media Indicator', () => {
  shouldShallowMatchSnapshot(
    'should render an audio item correctly',
    <MediaIndicator item={audioItem} service="news" />,
  );

  shouldShallowMatchSnapshot(
    'should render a video item correctly',
    <MediaIndicator item={videoItem} service="news" />,
  );

  shouldShallowMatchSnapshot(
    'should render a photo gallery item correctly',
    <MediaIndicator item={photogalleryItem} service="news" />,
  );

  shouldShallowMatchSnapshot(
    'should render correctly even without duration',
    <MediaIndicator item={noDurationItem} service="news" />,
  );

  shouldShallowMatchSnapshot(
    'should not render if item media object has no format',
    <MediaIndicator item={noMediaFormat} service="news" />,
  );

  isNull(
    'should not render if item is not media',
    <MediaIndicator item={nonMediaItem} service="news" />,
  );

  isNull(
    'should not render if item media object is missing',
    <MediaIndicator item={noMediaObject} service="news" />,
  );
});
