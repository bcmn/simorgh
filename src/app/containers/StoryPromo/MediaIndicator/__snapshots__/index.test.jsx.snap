// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story Promo Media Indicator should not render if item media object has no format 1`] = `
<MediaIndicator
  datetime="PT1H33M20S"
  duration="1:33:20"
  service="news"
  topStory={false}
  type={null}
/>
`;

exports[`Story Promo Media Indicator should render a photo gallery item correctly 1`] = `
<MediaIndicator
  datetime={null}
  duration={null}
  service="news"
  topStory={false}
  type="photogallery"
/>
`;

exports[`Story Promo Media Indicator should render a video item correctly 1`] = `
<MediaIndicator
  datetime="PT1H33M20S"
  duration="1:33:20"
  service="news"
  topStory={false}
  type="video"
/>
`;

exports[`Story Promo Media Indicator should render an audio item correctly 1`] = `
<MediaIndicator
  datetime="PT59S"
  duration="0:59"
  service="news"
  topStory={false}
  type="audio"
/>
`;

exports[`Story Promo Media Indicator should render correctly even without duration 1`] = `
<MediaIndicator
  datetime={null}
  duration={null}
  service="news"
  topStory={false}
  type="video"
/>
`;
